R/make_split_fun.R

```diff
@@ -0,0 +1,294 @@
+        stop("Split result element ", nm, " does not have required length ", len, ".",
+             msg_2_append)
+    TRUE
+}
+
+validate_split_result <- function(pinfo, component = NULL) {
+    assert_splres_element(pinfo, "datasplit", component = component)
+    len <- length(pinfo$datasplit)
+    assert_splres_element(pinfo, "values", len, component = component)
+    assert_splres_element(pinfo, "labels", len, component = component)
+    TRUE
+}
+
+#' Construct split result object
+#' These functions can be used to create or add to a split result in functions which
+#' implement core splitting or post-processing within a custom split function.
+#'
+#' @param values character or list<SplitValue>. The values associated with each facet
+#' @param datasplit list<data.frame>. The facet data for each facet generated in the split
+#' @param labels character. The labels associated with each facet
+#' @param extras NULL or list. Extra values associated with each of the facets which will be passed to analysis functions applied within the facet.
+#'
+#' @return a named list representing the facets generated by the split
+#'     with elements  `values`, `datasplit`,  and `labels`,  which are
+#'     the same length and correspond to eachother elementwise.
+#'
+#' @details
+#' These functions does various housekeeping to ensure that the split result
+#' list is as the rtables internals expect it, most of which are not
+#' relevant to end users.
+#'
+#'
+#' @examples
+#' splres <- make_split_result(values = c("hi", "lo"), datasplit = list(hi = mtcars, lo = mtcars[1:10,]), labels = c("more data", "less data"))
+#'
+#' splres2 <- add_to_split_result(splres, values = "med", datasplit = list(med = mtcars[1:20,]), labels = "kinda some data")
+#' @rdname make_split_result
+#' @export
+#' @family make_custom_split
+make_split_result <- function(values, datasplit, labels, extras = NULL) {
+    if(length(values) == 1 && is(datasplit, "data.frame"))
+        datasplit <- list(datasplit)
+    ret <- list(values = values, datasplit = datasplit, labels = labels)
+    if(!is.null(extras))
+        ret$extras <- extras
+    .fixupvals(ret)
+}
+
+#' @rdname make_split_result
+#' @param splres list. A list representing the result of splitting.
+#' @export
+add_to_split_result <- function(splres, values, datasplit, labels, extras = NULL) {
+    validate_split_result(splres)
+    newstuff <- make_split_result(values, datasplit, labels, extras)
+    ret <- lapply(names(splres),
+                  function(nm) c(splres[[nm]], newstuff[[nm]]))
+    names(ret) <- names(splres)
+    .fixupvals(ret)
+}
+
+.can_take_spl_context <- function(f) any(c(".spl_context", "...") %in% names(formals(f)))
+
+#' Create a Custom Splitting Function
+#'
+#' @param pre list. Zero or more functions which operate on the incoming data and return a new data frame that should split via `core_split`. They will be called on the data in the order they appear in the list
```

Comment:
so the output of the previous function will be the `df` argument of the next function on the list?

Reply:
@wwojciech correct, each `pre` function modifies `df` and provides the input for the next one in the list.